validator/src/journal/candidate_block.rs

```diff
@@ -220,4 +227,180 @@ impl CandidateBlock {
             .call_method(py, "set_signature", (signature,), None)
             .expect("BlockBuilder has no method 'set_signature'");
     }
+
+    pub fn finalize(&mut self, force: bool) -> Result<FinalizeBlockResult, BlockPublisherError> {
+        if !force || self.pending_batches.is_empty() {
+            return Err(BlockPublisherError::NoPendingBatchesRemaining);
+        }
+
+        if {
+            let py = unsafe { cpython::Python::assume_gil_acquired() };
+            !self.consensus
+                .call_method(
+                    py,
+                    "check_publish_block",
+                    (self.block_builder
+                        .getattr(py, "block_header")
+                        .expect("BlockBuilder has no attribute 'block_header'"),),
+                    None,
+                )
+                .expect("consensus has no method 'check_publish_block'")
+                .extract::<bool>(py)
+                .unwrap()
+        } {
+            return Err(BlockPublisherError::ConsensusNotReady);
+        }
+
+        self.scheduler.finalize(true).unwrap();
+        let execution_results = self.scheduler.complete(true).unwrap().unwrap();
+
+        let mut committed_txn_cache = {
+            let py = unsafe { cpython::Python::assume_gil_acquired() };
+            TransactionCommitCache::new(self.block_store.clone_ref(py))
+        };
+
+        let batches_w_no_results: Vec<String> = execution_results
+            .batch_results
+            .iter()
+            .filter(|(_, txns)| txns.is_none())
+            .map(|(batch_id, _)| batch_id.clone())
+            .collect();
+
+        let valid_batch_ids: Vec<String> = execution_results
+            .batch_results
+            .into_iter()
+            .filter(|(_, txns)| match txns {
+                Some(t) => !t.iter().any(|t| !t.is_valid),
+                None => false,
+            })
+            .map(|(b_id, _)| b_id)
+            .collect();
+
+        let builder = {
+            let py = unsafe { cpython::Python::assume_gil_acquired() };
```

Comment:
We should try to only call this once at the beginning of the function if it is needed because the GIL is already in scope for the entire function.